[Event Hubs Client] Track One: Ephemeral Test Resources

jsquire opened the pull request:
# Summary

The intent of these changes is to revise the approach used for dynamic resource management by the test suite, shifting to an approach where an ephemeral Event Hubs namespace and Azure Storage account are created upon request by a Live test and remains only for the duration of a given test run.

### Goals

- Reduce contention on a single Event Hubs namespace when tests are run; full parallelization of the track one and track two tests suites for each of their associated platforms  has begun to overwhelm the management plane with requests against a single namespace.  

- Reduce Azure spend for the cost center by limiting the time that charges are accrued by having an active resource only when being used.   This also allows for a greater number of throughput units to be allocated without increasing spend.

- Tune retry configuration to account for additional contention on Azure resources due to parallelization.

- Revise the associated assets and documentation to align with the changes.

# Last Upstream Rebase

Thursday, August 15, 2019  1:34pm (EDT)

jsquire committed:
[Event Hubs Client] Track One: Ephemeral Test Namespaces

- Enhanced resource management approach for tests moving to a strategy where each
  test run creates an Event Hubs namespace instance under which individual tests
  continue to use dedicated Event Hub instances.  Creation is only performed when
  an Event Hub scope is requested by a Live test.

- Tune retry configuration to account for additional contention on Azure resources
  due toe parallelization.

- Update ReadMe to enhance the "running tests" instructions to match the current
  implementation.

- Replaced the ARM template with a PowerShell script that will create the persistent
  resources needed for tests.

conniey commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,383 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Is there a unified way to do this across different client libraries running live services? I remember reading a thread in Teams, but curious why Powershell over ARM templates?

jsquire replied:
Service Principals and permissions are not, as far as I know or was able to determine, available in ARM.... so, Powershell.  

There's been a little bit of chatter about centralizing, but I don't think the approach or concept is fully baked yet and was mostly centered around the more standard needs of services using recorded tests with the HTTP pipeline, at present.

jsquire committed:
[Event Hubs Client] Track One: Ephemeral Test Storage Accounts

- Enhanced resource management approach for tests moving to a strategy where each
  test run creates an Azure storage instance that is removed when the run completes.
  Creation is only performed when an Event Hub scope is requested by a Live test.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+  Write-Host "$($indent)for use with the Event Hubs client library Live test suite."
+  Write-Host ""
+  Write-Host "$($indent)Upon completion, the script will output a set of environment variables with sensitive information which"
+  Write-Host "$($indent)fare used for testing.  When running Live tests, please be sure to have these environment variables available,"
```
Typo: fare => are

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+  Write-Host ""
+
+  Write-Host "$($indent)-ServicePrincipalName`tThe name to use for the service principal that will be"
+  Write-Host "$($indent)`t`t`t`tbe created to manage the Event Hub instances dynamically for the tests.  This"
```
nit: repeated word. "that will be **be** created".

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+{
+    <#
+      .SYNOPSIS
+        Cleans up any Azure resoruces created by the script.
```
Typo: resources.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+        Cleans up any Azure resoruces created by the script.
+        
+      .DESCRIPTION
+        Responsible for cleaning up any Azure resoruces created 
```
Typo: resources.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+    $AzureRegion = "southcentralus"
+}
+
+# Disallow prinicpal names with a space.
```
Typo: principal.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+
+if ($createResourceGroup)
+{
+    Write-Host "`t...Creaating new resource group"
```
Typo: Creating.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+
+if ($resourceGroup -eq $null)
+{
+    Write-Error "Unable to locate or create the resource goup: $($ResourceGroupName)"
```
Type: group.

kinelski replied:
I mean, "Typo". But I'll leave it here for the sake of irony.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+}
+
+# At this point, we may have created a resource, so be safe and allow for removing any
+# resoruces created should the script fail.
```
Typo: resources.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+
+try 
+{
+    # Create the service prinicpal and grant contributor access for management in the resource group.
```
Typo: principal.

kinelski commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+        New-AzRoleAssignment -ApplicationId "$($principal.ApplicationId)" -RoleDefinitionName "Contributor" -ResourceGroupName "$($ResourceGroupName)" | Out-Null
+    }    
+
+    # Write the environment variables
```
Super nit: maybe we ending the comment with a period to keep it consistent with other comments in the file?

kinelski reviewed (approved):
I'm not a PowerShell expert, but it LGTM. Added a few comments regarding typos and small stuff.

jsquire commented:
Obvious that the spelling checker does not follow me to where I wrote the PowerShell, eh?   :)

conniey commented on sdk/eventhub/Azure.Messaging.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -254,22 +254,22 @@ if ($resourceGroup -eq $null)
 if ($resourceGroup -eq $null)
 {
-    Write-Error "Unable to locate or create the resource goup: $($ResourceGroupName)"
+    Write-Error "Unable to locate or create the resource group: $($ResourceGroupName)"
```
I was hoping you'd edit it to `gout`

jsquire replied:
![](https://media3.giphy.com/media/t6KmBEEzokJ8CS5bBq/giphy.gif)

conniey commented on sdk/eventhub/Microsoft.Azure.EventHubs/assets/live-tests-azure-setup.ps1:
```diff
@@ -0,0 +1,325 @@
+)
+
+# =====================
+# == Module Imports  ==
```
```suggestion
# == Module Imports ==
```

conniey commented on sdk/eventhub/Microsoft.Azure.EventHubs/tests/Amqp/AmqpMEssageCoverterTests.cs:
```diff
@@ -55,7 +55,8 @@ private static ArraySegment<byte>[] ReadMessagePayLoad(AmqpMessage message, int
                 {
-                    foreach (var segment in messageBuffers) { message.CompletePayload(segment.Count); }
+                    foreach (var segment in messageBuffers)
+                    { message.CompletePayload(segment.Count); }
```
Would it be more .NET to put it on separate lines?
```cs
{
    message.CompletePayload(segment.Count);
}
```

jsquire replied:
Without a doubt;  however, this is track one code, so I'm making just the minimal set of changes.  :)

jsquire committed:
Apply suggestions from code review

Co-Authored-By: Connie Yau <conniey@microsoft.com>
